sonarr: 3.0.10.1567 -> 4.0.0.748

purcell opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

As pointed out in #278002 we were lagging a long way behind the latest releases, with v4 coming out in Nov 2022: https://forums.sonarr.tv/t/sonarr-v4-beta/31078

There's no longer a plain "linux" (actually cross-platform) package available for download, so I made an attempt to build from source (like jellyfin) but abandoned it due to obscure issues with an apparently-vendored .NET library, instead opting to follow the pattern of per-platform downloads used for `radarr`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

purcell committed:
sonarr: 3.0.10.1567 -> 4.0.0.748

purcell commented:
Perhaps you'd like to give this a try, @dacioromero.

purcell commented:
Open questions from my side:
- The 4.x release claims to not need mono, so I tried switching the wrapper script to `dotnet-runtime` like `radarr`, but didn't get it working, and in the meantime the program appears to launch find as-is
- Unsure if the on-disk config is compatible, and whether it's therefore potentially necessary to tag the config version

justinas commented on pkgs/servers/sonarr/default.nix:
```diff
@@ -1,12 +1,28 @@
+    arm64-linux_hash = "sha256-4jQzgjicPh4AmO2jY3h8YXlTlvIbk4uyDxMpOQtm+HI=";
+    x64-osx_hash = "sha256-MYLbAg9oCjWaZEMnSBmLEDEpab9vvcJjI8P8Fdmhmzs=";
+    arm64-osx_hash = "sha256-FyHffgNVoS/HPXz2YQGyxRowT04VFKtWZYEvsBZ9t4E=";
+  }."${arch}-${os}_hash";
```
nit: the `_hash` suffix seems superfluous, why not make the keys just `${arch}-${os}`.

purcell replied:
Sure, maybe, just copying the pattern used in `radarr` here. If you look in the update script, it at least makes the line unambiguously grep-able.

justinas replied:
Ah, makes sense, disregard my suggestion :slightly_smiling_face:

justinas commented on pkgs/servers/sonarr/default.nix:
```diff
@@ -1,12 +1,28 @@
-    url = "https://download.sonarr.tv/v3/main/${version}/Sonarr.main.${version}.linux.tar.gz";
-    hash = "sha256-6zdp/Bg+9pcrElW5neB+BC16Vn1VhTjhMRRIxGrKhxc=";
+    url = "https://download.sonarr.tv/v4/main/${version}/Sonarr.main.${version}.${os}-${arch}.tar.gz";
+    sha256 = hash;
```
nit: `inherit hash` instead?

purcell replied:
Again, copying the `radarr` case, but I'd generally agree.

justinas commented:
[Stable release announcement](https://forums.sonarr.tv/t/sonarr-v4-released/33089).

Some points of note:

> ffprobe has replaced MediaInfo - no more crashes from failed media scans

We probably need to add `ffmpeg` to sonarr's `PATH`?

> Due to multiple database migrations we’ve seen that some corrupt databases that were doing OK in v3 have broken in v4.

We should probably need to at least note the possible breakage in the 24.05 release notes.

purcell commented:
I've force-pushed a couple of changes trying to get this to run on my `aarch64-linux` machine, but no success so far. Launching with `mono` reported that the executable was not a valid assembly, but launching the `Sonarr` executable directly doesn't work either because it's dynamically linked:
```
Jan 01 16:23:24 media NzbDrone[397086]: Could not start dynamically linked executable: /nix/store/kppfskw8fwm6gsdhj594lla9vnnmnaii-sonarr-4.0.0.748/bin/Sonarr
Jan 01 16:23:24 media NzbDrone[397086]: NixOS cannot run dynamically linked executables intended for generic
Jan 01 16:23:24 media NzbDrone[397086]: linux environments out of the box. For more information, see:
Jan 01 16:23:24 media NzbDrone[397086]: https://nix.dev/permalink/stub-ld
```
In an ideal world we'd just build from source, but that started to feel painful. I pushed my changes [to a branch](https://github.com/NixOS/nixpkgs/compare/master...purcell:sonarr-build-from-source) in case anyone is interested.

purcell commented:
I think I have this working now, using the `dotnet-runtime` to run the DLL, and adding the necessary runtime dependencies (`ffmpeg`, as suggested, plus `icu` and `openssl`). And I applied the `inherit hash` suggestion too. So this feels in reasonable shape to me now.

I also added a release note to flag the potential config issue, and rebased against `master`.

purcell commented:
One thing I decided not to do here was renaming the main executable we produce from `NzbDrone` to `sonarr` — the (historical) NzbDrone name is still used in the config path in the service definition, and I didn't want to risk breaking that.

dacioromero commented:
@purcell Thanks for getting on this! I just tried importing my v4 backup from an OCI container I was running and it seems to have mostly worked except it failed to restart. Restarting the service manually seems to fix it.

```
Jan 01 17:12:21 fiyarr NzbDrone[74440]: [Info] LifecycleService: Restart requested.
Jan 01 17:12:21 fiyarr NzbDrone[74543]: The command could not be loaded, possibly because:
Jan 01 17:12:21 fiyarr NzbDrone[74543]:   * You intended to execute a .NET application:
Jan 01 17:12:21 fiyarr NzbDrone[74543]:       The application '/data=/var/lib/sonarr/.config/NzbDrone' does not exist.
Jan 01 17:12:21 fiyarr NzbDrone[74543]:   * You intended to execute a .NET SDK command:
Jan 01 17:12:21 fiyarr NzbDrone[74543]:       No .NET SDKs were found.
Jan 01 17:12:21 fiyarr NzbDrone[74543]: Download a .NET SDK:
Jan 01 17:12:21 fiyarr NzbDrone[74543]: https://aka.ms/dotnet-download
Jan 01 17:12:21 fiyarr NzbDrone[74543]: Learn about SDK resolution:
Jan 01 17:12:21 fiyarr NzbDrone[74543]: https://aka.ms/dotnet/sdk-not-found
Jan 01 17:12:21 fiyarr systemd[1]: sonarr.service: Deactivated successfully.
Jan 01 17:12:21 fiyarr systemd[1]: sonarr.service: Consumed 5.630s CPU time, received 3.3M IP traffic, sent 69.5K IP traffic.
```

Migrating from v3 to v4 also seems to work. I reverted to v3, restored my v3 backup, and then upgraded and all my series were still there.

This is how I am testing v4

```nix
  services.sonarr =
    let
      sonarr4 = import
        (builtins.fetchTarball {
          url = "https://github.com/purcell/nixpkgs/archive/16d725c6f48995ed905d0f70adf7f9a9a4445415.tar.gz";
          sha256 = "02daj2xgq9k9mp7s3xdh7p6m0i4l8i3id7i7kxzgpfh9r8h6h3jl";
        })
        {
          inherit (config.nixpkgs) config;
          system = "x86_64-linux";
        };
      package = sonarr4.sonarr;
    in
    {
      enable = true;
      openFirewall = true;
      user = "media";
      group = "media";
      inherit package;
    };
```

purcell commented:
Thanks for the extra data point, @dacioromero.

Seems to still be happily running for me without issues, so I think this is ready for review and potential merge.

purcell commented:
Also pinging @edwtjo, who might be well placed to quickly review this: we both maintain the `radarr` package here.

Sorixelle commented:
Heya, just noticed this PR :wave: In case you're interested, I have a working derivation building Sonarr v4 from source that I threw together for my own use while it was in beta. It's a little rough around the edges, but I figured I'd drop it in here in case it's helpful for the maintainers working on this PR: https://gist.github.com/Sorixelle/0d577d5e8632105be709a643352c0d88

purcell commented:
@Sorixelle That's great, thanks so much for sharing! My suggestion would be that we merge this PR first, and then as one of the maintainers I'll definitely pursue getting the source build working based on your code.

purcell commented:
Resolved conflict in the changelog, still ready for merge.

tie commented:
Since Sonarr v4 should be a drop-in replacement for Sonarr v3, perhaps we can keep the current v3 `sonarr` package and instead introduce `sonarr_4` with backport to 23.11 release? And for 24.05, rename `sonarr_4` back to `sonarr`. This would allow upgrading to Sonarr v4 before 24.05 release, e.g. by setting `services.sonarr.package = pkgs.sonarr_4` for people on the stable 23.11 release.

>My suggestion would be that we merge this PR first, and then as one of the maintainers I'll definitely pursue getting the source build working based on your code.

FWIW here is my take on source build: https://github.com/tie-infra/nix-config/tree/5db5a22c753c4b15549c750f75802c0049ab5671/parts/sonarr-v4
It needs an update script and there are two minor workarounds for issues in the underlying Nix infrastructure for .NET, but those should be trivial to fix.

tie commented on pkgs/servers/sonarr/default.nix:
```diff
@@ -1,25 +1,42 @@
+      --add-flags "$out/share/${pname}-${version}/Sonarr.dll" \
       --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
-          curl sqlite libmediainfo ]}
+          curl sqlite openssl ffmpeg icu ]}
```
>We probably need to add ffmpeg to sonarr's PATH?

There should be a precompiled statically-linked `ffprobe` binary bundled, see also https://github.com/Servarr/FFMpegCore/tree/master#supporting-both-32-and-64-bit-processes

Since it’s a `PATH` lookup, I don’t think adding ffmpeg to `LD_LIBRARY_PATH` makes sense.

purcell replied:
For comparison, Jellyfin bundles an ffmpeg binary but we still include the one from nixpkgs. Agree re. path vs. LD_LIBRARY_PATH, so I've moved that item to `--prefix PATH`.

tie replied:
I’m not really sure if Sonarr prefers `ffprobe` from `PATH` to the bundled binary. Regardless, we should probably also remove it from the package, that is, `rm ffprobe` before copying to `$out`.

SuperSandro2000 replied:
Lets move this into a seperate PR as this could cause unforeseen regressions.

purcell commented:
> FWIW here is my take on source build

Thanks @tie: using your code the build failed for me locally (aarch64-darwin, sandboxed). 😔   https://gist.github.com/purcell/62d4bb42681e560a39035454f9f04243

purcell commented:
> Since Sonarr v4 should be a drop-in replacement for Sonarr v3, perhaps we can keep the current v3 sonarr package and instead introduce sonarr_4 with backport to 23.11 release?

Maybe? It's probably not for me to say, as a somewhat casual maintainer.

tie commented:
>using your code the build failed for me locally (aarch64-darwin, sandboxed). 😔

Uh, I think it would be easier to disable checks on Darwin, https://github.com/tie-infra/nix-config/commit/b94250d93effd60d1832cfff2b06e1d36b3f4f6b
Builds fine for me on `aarch64-darwin` after this change.

tie commented:
>Maybe? It's probably not for me to say, as a somewhat casual maintainer. It sounds reasonable but I am unfamiliar with the processes, conventions and logistics needed to make it happen.

In general, as long as the change is [acceptable for releases (as in CONTRIBUTING.md)](https://github.com/NixOS/nixpkgs/blob/23.11/CONTRIBUTING.md#changes-acceptable-for-releases), tagging the PR with [backport release-YY.MM label](https://github.com/NixOS/nixpkgs/labels?q=backport) is enough, and the bot then creates backport PR (that is merged separately). In this case, adding `sonarr_4` package while keeping v3 as `sonarr` (and potentially `sonarr_3`) wouldn’t be a breaking change.

---

Regarding source build, I took some time to implement fixes for `buildDotnetModule` and I don’t think we can reasonably backport those, and I wouldn’t feel comfortable backporting a package with workarounds (or a package that can’t be cross-compiled). See also https://github.com/NixOS/nixpkgs/compare/staging...tie:nixpkgs:sonarr-v4

purcell commented:
> In general, as long as the change is [acceptable for releases (as in CONTRIBUTING.md)](https://github.com/NixOS/nixpkgs/blob/23.11/CONTRIBUTING.md#changes-acceptable-for-releases), tagging the PR with [backport release-YY.MM label](https://github.com/NixOS/nixpkgs/labels?q=backport) is enough, and the bot then creates backport PR (that is merged separately). In this case, adding `sonarr_4` package while keeping v3 as `sonarr` (and potentially `sonarr_3`) wouldn’t be a breaking change.

This would be the first time such a process has been used for either `sonarr` or its sibling `radarr`, and I worry that we're overcomplicating things by suggesting it tbh. End-user functionality and config for this release are more or less unchanged, so it probably qualifies for release _broadly_ as there are no intentional breaking changes.

The only reason folks are suggesting the `_3` -> `_4` process is the report of broken automatic config migrations, but this is not something that end users can really predict or adjust for, and it apparently only affected a small minority of people with very old configs that had been incrementally upgraded over multiple versions.

If the latter is sufficient to be considered a breaking change, and the release note isn't enough, I would personally suggest we not backport this rather than do the `_3` -> `_4` dance, as the existing release version still works fine despite being a year out of date.

purcell commented:
How to move this forward? I'm a maintainer for this derivation, but I don't have the commit bit. What buy-in is necessary for the change to get merged in its current form, or who decides what _must_ or _should_ be done here?

eyJhb commented:
I've been using this patch for some time now, and I don't think we should wait to merge this. Generally the move from mono to dotnet provides many benefits, one of them (and the reason I've started to use this patch), is because of IPv6 support.

Adding it to the release notes should suffice, but, I'm not sure if we can get any higher power to look into this.

mweinelt commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -130,6 +130,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 - The option [`services.nextcloud.config.dbport`] of the Nextcloud module was removed to match upstream.
   The port can be specified in [`services.nextcloud.config.dbhost`](#opt-services.nextcloud.config.dbhost).
 
+- With a bump to `sonarr` v4, existing config database files will be upgraded automatically, but the [upstream release announcement](https://forums.sonarr.tv/t/sonarr-v4-released/33089) reports that such upgrades have failed in the past when existing working configs were actually corrupted.
```
How is this actionable to the user?

purcell replied:
It's really not actionable, other than a user choosing not to upgrade because they're worried, which I can't really imagine.

Perhaps this is better? Or we can just omit, it's not a big deal as far as I'm concerned.
```suggestion
- With a bump to `sonarr` v4, existing config database files will be upgraded automatically, but note that some old quietly-corrupted configs [might fail to upgrade cleanly](https://forums.sonarr.tv/t/sonarr-v4-released/33089).
```

purcell commented:
Rebased to resolve conflict in release notes.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3389

SuperSandro2000 commented on pkgs/development/libraries/linenoise/default.nix:
```diff
@@ -1,29 +1,48 @@
     repo = "linenoise";
-    rev = "c894b9e59f02203dbe4e2be657572cf88c4230c3";
-    sha256 = "0wasql7ph5g473zxhc2z47z3pjp42q0dsn4gpijwzbxawid71b4w";
+    rev = "1.0-34-g93b2db9";
```
```suggestion
    rev = version;
```

purcell replied:
Thanks, this was a modification in an unrelated file that crept in via a rebase. Fixed now.

SuperSandro2000 commented on pkgs/servers/sonarr/default.nix:
```diff
@@ -1,25 +1,43 @@
-    cp -r * $out/bin/
-    makeWrapper "${mono}/bin/mono" $out/bin/NzbDrone \
-      --add-flags "$out/bin/Sonarr.exe" \
+    mkdir -p $out/{bin,share/${pname}-${version}}
```
```suggestion
    mkdir -p $out/{bin,share/sonarr-${version}}
```
pname should not be lightly substituted

SuperSandro2000 commented on pkgs/servers/sonarr/default.nix:
```diff
@@ -1,25 +1,43 @@
-    makeWrapper "${mono}/bin/mono" $out/bin/NzbDrone \
-      --add-flags "$out/bin/Sonarr.exe" \
+    mkdir -p $out/{bin,share/${pname}-${version}}
+    cp -r * $out/share/${pname}-${version}/.
```
```suggestion
    cp -r * $out/share/sonarr-${version}/.
```

SuperSandro2000 commented on pkgs/servers/sonarr/default.nix:
```diff
@@ -1,25 +1,43 @@
+    cp -r * $out/share/${pname}-${version}/.
+
+    makeWrapper "${dotnet-runtime}/bin/dotnet" $out/bin/NzbDrone \
+      --add-flags "$out/share/${pname}-${version}/Sonarr.dll" \
```
```suggestion
      --add-flags "$out/share/sonarr-${version}/Sonarr.dll" \
```

SuperSandro2000 commented on pkgs/servers/sonarr/default.nix:
```diff
@@ -1,25 +1,43 @@
+      --add-flags "$out/share/${pname}-${version}/Sonarr.dll" \
+      --prefix PATH : ${lib.makeBinPath [ ffmpeg ]} \
       --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
-          curl sqlite libmediainfo ]}
+          curl sqlite openssl icu ]}
```
```suggestion
      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ curl sqlite openssl icu ]}
```
nit

SuperSandro2000 reviewed (commented):
especially the version re-use must be fixed before this can be merged

purcell committed:
sonarr: 4.0.0.748 -> 4.0.1.929

Also switch to using assets attached to GitHub releases, which now
seems to be the official download location.

SuperSandro2000 reviewed (approved):
diff LGTM, didn't try myself but I want to free the author out of his waiting 🙈
